### ISSUE TITLE
"Failed to verify signature" but SPSSODescriptor validation is disabled

### ISSUE THREAD

**Issue by GensaGames** [latchset/mod_auth_mellon#119]
When providing metadata to the IdP with disabled signature verification: 

```
<EntityDescriptor entityID="https://xxxxxxxx/myEntityID" xmlns="urn:oasis:names:tc:SAML:2.0:metadata" xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
        <SPSSODescriptor
                protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol"
                AuthnRequestsSigned="false"
                WantAssertionsSigned="false"
                >
....
```

IdP sign in succeed and I have ongoing post response to `/secret/postResponse` - it fails on signature verification, and I have 400 response. Any ideas what is wrong here? I don't see anything related on mod_auth_mellon to disable signature verification, but it still kicks in. I think it should pick up my metadata configuration where I have disabled validation. 

```
[Wed Mar 15 05:35:12.189284 2023] [auth_mellon:error] [pid 1444326:tid 140279633172224] [client xxx.31.11.91:16588] Error processing authn response. Lasso error: [-111] Failed to verify signature., SAML Response: StatusCode1="urn:oasis:names:tc:SAML:2.0:status:Success", StatusCode2="(null)", StatusMessage="(null)", referer: xxx
```

This is how Apache config looks like on the machine. Note that application is hosted on HTTP and SSL certificates settings and secure handling on an ALB. 

```
  <VirtualHost *:80>
    ProxyPreserveHost On
    ProxyRequests Off
    ProxyPass / http://localhost:8088/
    ProxyPassReverse / http://localhost:8088/
    ProxyPreserveHost On
    AllowEncodedSlashes On

    ## Enable the mellon plugin and trigger auth to the IdP
    <Location />

            Require valid-user
            AuthType "Mellon"
            MellonEndpointPath "/secret"
            MellonSamlResponseDump Off
            MellonSecureCookie Off
            MellonPostReplay Off

            MellonEnable "auth"
            MellonSPPrivateKeyFile /etc/apache2/mellon/xxxxxxxxxxx_myEntityID.key
            MellonSPCertFile /etc/apache2/mellon/xxxxxxxxxxx_myEntityID.cert
            MellonSPMetadataFile /etc/apache2/mellon/xxxxxxxxxxx_myEntityID.xml
            MellonIdPMetadataFile /etc/apache2/mellon/federate_metadata_saml_202303_20230311.xml
        </Location>

  </VirtualHost>
```

**Comment by GensaGames** [comment]
On this note, I have interesting problem. When I was trying to change IdP metadata I have removed `KeyDescriptor` with IdP CA inside.  At the same time I specified `MellonIdPPublicKeyFile` inside Apache config. Based on documentation: 

> 
>         # MellonIdpPublicKeyFile is the full path of the public key of the
>         # IdP. This parameter is optional if the public key is embedded
>         # in the IdP's metadata file, or if a certificate authority is
>         # used. This parameter cannot be used if multiple IdP are configured.
>         # Default: None set.
>         MellonIdPPublicKeyFile /etc/apache2/mellon/idp-public-key.pem


Regardless of this option I was getting the same error about missing CA inside my metadata. 

```
Apache-Error: [file "auth_mellon_handler.c"] [line 2134] [level 3] Error processing authn response. Lasso error: [-1500] The provider has no known public key, SAML Response: StatusCode1="urn:oasis:names:tc:SAML:2.0:status:Success", StatusCode2="(null)", StatusMessage="(null)"
Apache-Handler: proxy-server
```

When I place CA inside the xml, it's coming back normal (I mean with another error which I have above). At this point, I think there is some problem with loading these CA from configuration and metadata for validation signature (or verifying public CA of the Idp).

**Comment by thijskh** [comment]
* Metadata is not the way to configure Mellon; metadata is what you provide to the remote party.
WantAssertionsSigned="false" is therefore an instruction to the IdP then what it should send, not what the receiving SP would do.

* WantAssertionSigned does not mean to completely disable all signatures, but means that the Assertion xml element should not be signed, the response element will then still be signed.

* There is no way to completely disable all signature validation since this will render the SP completely insecure and useless since then everyone in the world can authenticate as every identity they want.

* It might be a bug that `MellonIdPPublicKeyFile` does not work, but placing the key inside the metadat works for sure.

**Comment by GensaGames** [comment]
That's for coming back on this. That make sense. Interesting, that CA of the IdP is there, and my SP certificates are inside the metadata as well. No matter how I change configuration, I'm still getting the same error related to the `Error processing authn response. Lasso error: [-111] Failed to verify signature`. I will keep checking it, thanks anyway.

**Comment by thijskh** [comment]
The unpredictable behaviour of `MellonIdPPublicKeyFile` has been fixed in 1465590626cf0e3db0cf7b4ea8a315908b3cecbe.

**Comment by GensaGames** [comment]
Better late than never,  Thank you Sir!
